README.md
# angular2-chartjs
Chart.js component for Angular2+

## Installation
You can install angular2-chartjs by using npm.
```
npm install angular2-chartjs
```

## Usage
Add `ChartModule` to your module, eg.
```JavaScript
import { ChartModule } from 'angular2-chartjs';

@NgModule({
  imports: [ ChartModule ]
  // ...
})
export class AppModule {
}
```
And you can use selector `chart` in your template.

JavaScript
```JavaScript
type = 'line';
data = {
  labels: ["January", "February", "March", "April", "May", "June", "July"],
  datasets: [
    {
      label: "My First dataset",
      data: [65, 59, 80, 81, 56, 55, 40]
    }
  ]
};
options = {
  responsive: true,
  maintainAspectRatio: false
};
```
HTML
```HTML
<chart [type]="type" [data]="data" [options]="options"></chart>
```

## Options
### [type]
*string*, required  
Chart type.

### [data]
*object*, required  
To display data, the chart must be passed a data object that contains all of the information needed by the chart. [See](http://www.chartjs.org/docs/#chart-configuration-chart-data).

### [options]
*object*, optional  
To create a chart with configuration options, simply pass an object containing your configuration to the constructor. [See](http://www.chartjs.org/docs/#chart-configuration-creating-a-chart-with-options).

## Events
### (clickCanvas)
*Event*  
Return click event.

### (clickDataset)
*Array*  
Return the clicked dataset array.

### (clickElement)
*Array*  
Return the clicked element in array.

### (clickElements)
*Array*  
Return the clicked elements in array.

## Members
### chart
*Chart*  
Chart instance. You can call Chart.js methods via this member.

Example:
```JavaScript
chartComponent.chart.destroy();
```

## SystemJs
Add following settings
```JavaScript
{
  map: {
    'angular2-chartjs': 'npm:angular2-chartjs',
    'chart.js': 'npm:chart.js/dist/Chart.bundle.js'
  },
  packages: {
    'angular2-chartjs': {
      main: './dist/index.js',
      defaultExtension: 'js'
    }
  }
}
```

## FAQ
How to call Chart.js methods - [#7](https://github.com/emn178/angular2-chartjs/issues/7), [#2](https://github.com/emn178/angular2-chartjs/issues/2)

## License
The project is released under the [MIT license](http://www.opensource.org/licenses/MIT).

## Contact
The project's website is located at https://github.com/emn178/angular2-chartjs  
Author: Chen, Yi-Cyuan (emn178@gmail.com)


CHANGELOG.md
# Change Log

## v0.5.1 / 2018-05-18
### Improve
- performance

## v0.5.0 / 2018-03-17
### Added
- events `clickCanvas`, `clickDataset`, `clickElements` and `clickElement`.

## v0.4.1 / 2017-11-19
### Fixed
- angular 5 aot build.

## v0.4.0 / 2017-11-15
### Fixed
- `Responsive` seems to be broken when used with chart.js >= 2.7.0 #26
- angular 5 warning.

## v0.3.0 / 2017-08-16
### Changed
- change type and options will re-create chart.

## v0.2.0 / 2017-04-07
### Added
- angular 4 peer dependencies.

## v0.1.6 / 2016-11-16
### Fixed
- issuse #1 for Ionic2.

## v0.1.5 / 2016-10-24
### Added
- exposed chart instance.

## v0.1.4 / 2016-10-13
### Fixed
- dependencies problem in some cases.

## v0.1.3 / 2016-10-05
### Fixed
- typings path in package.json.

## v0.1.2 / 2016-10-04
### Added
- Data change detection.

## v0.1.1 / 2016-10-03
### Added
- First version implementation.


src/chart.component.ts
import { Component, Input, ElementRef, OnInit, OnChanges, SimpleChanges, Output, EventEmitter, NgZone } from '@angular/core';

declare var Chart;

@Component({
  selector: 'chart',
  template: '',
  styles: [':host { display: block; }']
})
export class ChartComponent implements OnInit, OnChanges {
  chart: any;

  @Input() type: string;
  @Input() data: any;
  @Input() options: any;
  @Output() clickCanvas = new EventEmitter();
  @Output() clickDataset = new EventEmitter();
  @Output() clickElements = new EventEmitter();
  @Output() clickElement = new EventEmitter();

  private canvas;

  constructor(private elementRef: ElementRef, private ngZone: NgZone) { }

  ngOnInit() {
    this.create();
  }

  ngOnChanges(changes: SimpleChanges) {
    if (this.chart) {
      if (changes['type'] || changes['options']) {
        this.create();
      } else if (changes['data']) {
        let currentValue = changes['data'].currentValue;
        ['datasets', 'labels', 'xLabels', 'yLabels'].forEach(property => {
          this.chart.data[property] = currentValue[property];
        })
        this.chart.update();
      }
    }
  }

  private create() {
    this.ngZone.runOutsideAngular(() => {
      if (this.canvas) {
        this.elementRef.nativeElement.removeChild(this.canvas);
      }
      this.canvas = document.createElement('canvas');
      this.elementRef.nativeElement.appendChild(this.canvas);
      this.chart = new Chart(this.canvas, {
        type: this.type,
        data: this.data,
        options: this.options
      });
      this.canvas.onclick = e => {
        this.ngZone.run(() => {
          this.clickCanvas.next(e);
          if (this.clickDataset.observers.length) {
            this.clickDataset.next(this.chart.getDatasetAtEvent(e));
          }
          if (this.clickElements.observers.length) {
            this.clickElements.next(this.chart.getElementsAtEvent(e));
          }
          if (this.clickElement.observers.length) {
            this.clickElement.next(this.chart.getElementAtEvent(e));
          }
        });
      };
    });
  }
}


dev/app.component.ts
import { Component, OnInit, ViewChild } from '@angular/core';
import { ChartComponent } from '../src';

@Component({
  selector: 'app',
  template: '<button (click)="change()">Change</button><chart [type]="type" [data]="data" [options]="options"></chart>'
})
export class AppComponent {
  @ViewChild(ChartComponent) chartComponent: ChartComponent;

  type = 'line';
  data = {
    labels: ['January', 'February', 'March', 'April', 'May', 'June', 'July'],
    datasets: [
      {
        label: 'My First dataset',
        data: [65, 59, 80, 81, 56, 55, 40],
        backgroundColor: '#FF6384'
      },
      {
        label: 'My Second dataset',
        data: [60, 65, 75, 86, 49, 42, 35],
        backgroundColor: '#36A2EB'
      }
    ]
  };
  options = {
    responsive: true,
    maintainAspectRatio: false,
    scales: {
      xAxes: [
        {
          stacked: true
        }
      ],
      yAxes: [
        {
          stacked: true
        }
      ]
    }
  };

  constructor() {
  }

  change() {
    this.type = 'bar';
    this.data = {
      labels: ['January', 'February', 'March', 'April', 'May', 'June', 'July'],
      datasets: [
        {
          label: 'My First dataset',
          data: [65, 59, 80, 81, 56, 55, 40],
          backgroundColor: '#FF6384'
        },
        {
          label: 'My Second dataset',
          data: [10, 30, 20, 40, 60, 50, 70],
          backgroundColor: '#36A2EB'
        },
        {
          label: 'My Third dataset',
          data: [35, 59, 50, 61, 76, 46, 35],
          backgroundColor: '#FFCE56'
        }
      ]
    };
  }
}
